File: backend/app/utils/model_naming.py
"""
Centralized model naming utilities.

This module provides consistent naming logic for all model types
to ensure no hardcoded naming conventions across the codebase.
"""

def format_official_yolo_model_name(yolo_model: str) -> str:
    """
    Format official YOLO model name with consistent naming convention.
    
    Args:
        yolo_model: Model identifier (e.g., "yolov8n", "yolov9s", "yolo26n")
        
    Returns:
        Formatted model name following consistent pattern with space before scale
        
    Examples:
        - "yolov8n" -> "YOLOV8 N"
        - "yolov9s" -> "YOLOV9 S"
        - "yolo26n" -> "YOLO26 N"
        - "yolo11x" -> "YOLO11 X"
    """
    # Extract base name and scale letter
    if len(yolo_model) > 0 and yolo_model[-1].isalpha():
        base_name = yolo_model[:-1].upper()
        scale = yolo_model[-1].upper()
        return f"{base_name} {scale}"
    else:
        return yolo_model.upper()


def format_arch_plugin_model_name(arch_plugin, scale: str) -> str:
    """
    Format arch plugin model name with scale.
    
    Args:
        arch_plugin: Architecture plugin object
        scale: Scale character (e.g., "n", "s", "m", "l", "x")
        
    Returns:
        Formatted model name with scale
        
    Examples:
        - mamba_yolo + "n" -> "Mamba-YOLO N"
        - rtdetr + "l" -> "RT-DETR L"
        - yolo26_cs2ga + "n" -> "YOLO26 + CS²GA N"
    """
    base_name = arch_plugin.display_name
    if scale:
        return f"{base_name} {scale.upper()}"
    return base_name


def get_model_display_name(model_id: str, model_scale: str = None, arch_plugin=None) -> str:
    """
    Get standardized display name for any model.
    
    Args:
        model_id: Model identifier (e.g., "yolo:yolov8n", "arch:mamba_yolo")
        model_scale: Scale for arch plugins (e.g., "n", "s", "m")
        arch_plugin: Architecture plugin object for arch models
        
    Returns:
        Standardized model display name
    """
    if model_id.startswith("yolo:"):
        # Official YOLO model
        yolo_model = model_id.split(":")[1]
        return format_official_yolo_model_name(yolo_model)
    elif model_id.startswith("arch:") and arch_plugin:
        # Architecture plugin model
        return format_arch_plugin_model_name(arch_plugin, model_scale or "")
    else:
        # Custom model - use as-is or generate from model_id
        return model_id.replace("_", " ").replace("-", " ").title()

File: frontend/src/utils/modelNaming.ts
/**
 * Centralized model naming utilities for frontend.
 * 
 * This provides consistent naming logic across the frontend to match
 * the backend naming conventions.
 */

export function formatOfficialYoloModelName(yoloModel: string): string {
  /**
   * Format official YOLO model name with consistent naming convention.
   * 
   * Args:
   *   yoloModel: Model identifier (e.g., "yolov8n", "yolov9s", "yolo26n")
   * 
   * Returns:
   *   Formatted model name following consistent pattern with space before scale
   * 
   * Examples:
   *   - "yolov8n" -> "YOLOV8 N"
   *   - "yolov9s" -> "YOLOV9 S"
   *   - "yolo26n" -> "YOLO26 N"
   *   - "yolo11x" -> "YOLO11 X"
   */
  // Extract base name and scale letter
  if (yoloModel.length > 0 && /[a-zA-Z]/.test(yoloModel[yoloModel.length - 1])) {
    const baseName = yoloModel.slice(0, -1).toUpperCase();
    const scale = yoloModel[yoloModel.length - 1].toUpperCase();
    return `${baseName} ${scale}`;
  } else {
    return yoloModel.toUpperCase();
  }
}

export function formatArchPluginModelName(displayName: string, scale: string): string {
  /**
   * Format arch plugin model name with scale.
   * 
   * Args:
   *   displayName: Plugin display name (e.g., "Mamba-YOLO", "RT-DETR")
   *   scale: Scale character (e.g., "n", "s", "m", "l", "x")
   * 
   * Returns:
   *   Formatted model name with scale
   * 
   * Examples:
   *   - "Mamba-YOLO" + "n" -> "Mamba-YOLO N"
   *   - "RT-DETR" + "l" -> "RT-DETR L"
   *   - "YOLO26 + CS²GA" + "n" -> "YOLO26 + CS²GA N"
   */
  if (scale) {
    return `${displayName} ${scale.toUpperCase()}`;
  }
  return displayName;
}

export function getModelDisplayName(modelId: string, modelScale?: string, archDisplayName?: string): string {
  /**
   * Get standardized display name for any model.
   * 
   * Args:
   *   modelId: Model identifier (e.g., "yolo:yolov8n", "arch:mamba_yolo")
   *   modelScale: Scale for arch plugins (e.g., "n", "s", "m")
   *   archDisplayName: Display name for arch plugins
   * 
   * Returns:
   *   Standardized model display name
   */
  if (modelId.startsWith('yolo:')) {
    // Official YOLO model
    const yoloModel = modelId.split(':')[1];
    return formatOfficialYoloModelName(yoloModel);
  } else if (modelId.startsWith('arch:') && archDisplayName) {
    // Architecture plugin model
    return formatArchPluginModelName(archDisplayName, modelScale || '');
  } else {
    // Custom model - use as-is or generate from model_id
    return modelId.replace(/_/g, ' ').replace(/-/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
  }
}
